test/dsa_no_digest_size_test.c

```diff
@@ -0,0 +1,163 @@
+        TEST_error("DSA_generate_parameters_ex failure");
+        return 0;
+    }
+    if (!DSA_generate_key(dsa)) {
+        TEST_error("DSA_generate_key failed");
+        return 0;
+    }
+
+    return 1;
+}
+
+static int sign_and_verify(int len) {
+    // Per FIPS 186-4, the hash is recommended to be the same length as q.
+    // If the hash is longer than q, the leftmost N bits are used; if the hash
+    // is shorter, then we left-pad (see appendix C.2.1).
+
+    int digestlen = BN_num_bytes(DSA_get0_q(dsa));
+    int ok = 0;
+
+    unsigned char *dataToSign = malloc(len);
+    unsigned char *paddedData = malloc(digestlen);
+    unsigned char *signature = NULL;
+    EVP_PKEY_CTX *ctx = NULL;
+    EVP_PKEY *pkey = NULL;
+
+    if (!RAND_bytes(dataToSign, len)) {
+        TEST_error("RAND_bytes");
+        goto end;
+    }
+
+    memset(paddedData, 0, digestlen);
+    if (len > digestlen) {
+        memcpy(paddedData, dataToSign, digestlen);
+    } else {
+        memcpy(paddedData + digestlen - len, dataToSign, len);
+    }
+    
+    pkey = EVP_PKEY_new();
+    if (!pkey) {
+        TEST_error("EVP_PKEY_new");
+        goto end;
+    }
+    EVP_PKEY_set1_DSA(pkey, dsa);
+
+    ctx = EVP_PKEY_CTX_new(pkey, NULL);
+    if (!ctx) {
+        TEST_error("EVP_PKEY_CTX_new");
+        goto end;
+    }
+    if (EVP_PKEY_sign_init(ctx) <= 0) {
+        TEST_error("EVP_PKEY_sign_init");
+        goto end;
+    }
+
+    size_t sigLength;
+    int rv;
+    if ((rv = EVP_PKEY_sign(ctx, NULL, &sigLength, dataToSign, len)) <= 0) {
+        TEST_error("Failed to get signature length, len=%d", len);
+        goto end;
+    }
+
+    signature = malloc(sigLength);
+
+    if ((rv = EVP_PKEY_sign(ctx, signature, &sigLength, dataToSign, len)) <= 0) {
+        TEST_error("Failed to sign, len=%d", len);
+        goto end;
+    }
+
+    /* Check that the signature is okay via the EVP interface */
+    if (EVP_PKEY_verify_init(ctx) <= 0) {
+        TEST_error("Failed to switch to verify mode");
+        goto end;
+    }
+
+    /* ... using the same data we just signed */
+    if (EVP_PKEY_verify(ctx, signature, sigLength, dataToSign, len) != 1) {
+        TEST_error("EVP verify with unpadded length %d failed\n", len);
+        goto end;
+    }
+
+    /* ... padding/truncating the data to the appropriate digest size */
+    if (EVP_PKEY_verify(ctx, signature, sigLength, paddedData, digestlen) != 1) {
+        TEST_error("EVP verify with length %d failed\n", len);
+        goto end;
+    }
+
+    /* Verify again using the raw DSA interface */
+    if (DSA_verify(0, dataToSign, len, signature, sigLength, dsa) != 1) {
+        TEST_error("Verification with unpadded data failed, len=%d", len);
+        goto end;
+    }
+
+    if (DSA_verify(0, paddedData, digestlen, signature, sigLength, dsa) != 1) {
+        TEST_error("verify with length %d failed\n", len);
+        goto end;
+    }
+
+    ok = 1;
+end:
+    if (ctx) EVP_PKEY_CTX_free(ctx);
```

Comment:
Remove the non-NULL checks; the free routines ignore NULL.

Reply:
Whoops, you commented just as I was pushing a fix for some of the style issues. Hold off on looking at the update for a moment :)